packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,489 @@
+  },
+  allNonEvmTokens: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+// Define a temporary interface for the permission structure
+type AssetEndowment = {
+  'endowment:assets'?: {
+    caveats: CaveatConstraint[];
+  };
+};
+
+// TODO make this controller extends StaticIntervalPollingController and update all metadata once a day.
+
+export class MultichainAssetsController extends BaseController<
+  typeof controllerName,
+  MultichainAssetsControllerState,
+  MultichainAssetsControllerMessenger
+> {
+  // Mapping of CAIP-2 Chain ID to Asset Snaps.
+  #snaps: Record<CaipChainId, Snap[]>;
+
+  readonly #mutex = new Mutex();
+
+  constructor({
+    messenger,
+    state = {},
+  }: {
+    messenger: MultichainAssetsControllerMessenger;
+    state?: Partial<MultichainAssetsControllerState>;
+  }) {
+    super({
+      messenger,
+      name: controllerName,
+      metadata: assetsControllerMetadata,
+      state: {
+        ...getDefaultMultichainAssetsControllerState(),
+        ...state,
+      },
+    });
+
+    this.#snaps = {};
+
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAdded',
+      async (account) => await this.#handleOnAccountAdded(account),
+    );
+    this.messagingSystem.subscribe(
+      'AccountsController:accountRemoved',
+      async (account) => await this.#handleOnAccountRemoved(account),
+    );
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAssetListUpdated',
+      async (event) => await this.#updateAccountAssetsList(event),
+    );
+  }
+
+  /**
+   * Function to update the assets list for an account
+   *
+   * @param event - The list of assets to update
+   */
+  async #updateAccountAssetsList(event: AccountAssetListUpdatedEventPayload) {
+    const releaseLock = await this.#mutex.acquire();
+    try {
+      const assetsToUpdate = event.assets;
+      const assetsForMetadataRefresh: CaipAssetType[] = [];
+      for (const accountId in assetsToUpdate) {
+        if (hasProperty(assetsToUpdate, accountId)) {
+          const newAccountAssets = assetsToUpdate[accountId];
+          if (
+            newAccountAssets.added.length !== 0 ||
+            newAccountAssets.removed.length !== 0
+          ) {
+            const assets = this.state.allNonEvmTokens[accountId] || [];
+            const filteredAssetsToAdd = newAccountAssets.added.filter(
+              (asset) => !assets.includes(asset),
+            );
+            assetsForMetadataRefresh.push(...filteredAssetsToAdd);
+            const newAssets = [...assets, ...filteredAssetsToAdd];
+
+            const assetsAfterRemoval = newAssets.filter(
+              (asset) => !newAccountAssets.removed.includes(asset),
+            );
+            this.update((state) => {
+              state.allNonEvmTokens[accountId] = assetsAfterRemoval;
+            });
+          }
+        }
+      }
+      // trigger fetching metadata for new assets
+      await this.#refreshAssetsMetadata(assetsForMetadataRefresh);
```

Comment:
If we use the `Set` notation above, we could also make `assetsForMetadataRefresh` a `Set` too. (I would renamed it to `allAddedAssets` maybe).

This way we can just call `#refreshAssetsMetadata` with this set, and that will automatically refresh the metadata that are missing (since we're doing the filtering in this method, I guess that's ok to pass it "as-is").

WDYT?